File: Greedy Algorithms/MinimumWaitingTime.py
# My Solution:
# Time O(nlogn) | Space O(1) because we are sorting in place
def minimumWaitingTime(queries):
    # Write your code here.
    queries.sort()
	total_wait = 0
	cur_wait = 0 
	for i in range(1, len(queries)):
		cur_wait += queries[i-1]
		total_wait += cur_wait
		
	
	return total_wait


# Algo Expert Solution:
# Time O(nlogn) | Space O(1) - where n is the number of queries
def minimumWaitingTime(queries):
    queries.sort()
	
	total_wait = 0
	for idx, time in enumerate(queries):
		numQueriesLeft = len(queries) - (idx + 1)
		total_wait += time * numQueriesLeft
		
	
	return total_wait

File: Greedy Algorithms/MinimumWaitingTime.java
// Solution 2: The apprach used in the Algo Solution:
import java.util.*;

class Program {

  public int minimumWaitingTime(int[] queries) {
    Arrays.sort(queries);
		int curWaitTime = 0;
		int totalWaitTime = 0;
		for (int idx = 1; idx < queries.length; idx++) {
			curWaitTime += queries[idx-1];
			totalWaitTime += curWaitTime;
		}
    return totalWaitTime;
  }
}



// Solution 1: My intuitive solution
import java.util.*;

class Program {

  public int minimumWaitingTime(int[] queries) {
    Arrays.sort(queries);
		int curWaitTime = 0;
		int totalWaitTime = 0;
		for (int idx = 1; idx < queries.length; idx++) {
			curWaitTime += queries[idx-1];
			totalWaitTime += curWaitTime;
		}
		
			
		
    return totalWaitTime;
  }
}
